### ISSUE TITLE
GTK+ 3: panel main menu don't close properly

### ISSUE THREAD

**Issue by raveit65** [mate-desktop/mate-panel#305]
The panel freezes after adding a launcher form menu or desktop, in result most panel elements can not be executed until the panel is restarted by a 'killall mate-panel'.
Backtrace:
https://dl.dropboxusercontent.com/u/49862637/Mate-desktop/Bugs/gtk3_mate-panel_while_adding_launcher.txt
Also the panels spams the xsession.errors.log during every session start.

```
(mate-panel:1620): GdkPixbuf-CRITICAL **: gdk_pixbuf_get_height: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:1620): Gtk-CRITICAL **: gtk_widget_get_preferred_width_for_height: assertion 'height >= 0' failed

(mate-panel:1620): GdkPixbuf-CRITICAL **: gdk_pixbuf_get_height: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:1620): Gtk-CRITICAL **: gtk_widget_get_preferred_width_for_height: assertion 'height >= 0' failed

(mate-panel:1620): GdkPixbuf-CRITICAL **: gdk_pixbuf_get_height: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:1620): Gtk-CRITICAL **: gtk_widget_get_preferred_width_for_height: assertion 'height >= 0' failed

(mate-panel:1620): GdkPixbuf-CRITICAL **: gdk_pixbuf_get_height: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:1620): Gtk-CRITICAL **: gtk_widget_get_preferred_width_for_height: assertion 'height >= 0' failed

(mate-panel:1620): GdkPixbuf-CRITICAL **: gdk_pixbuf_get_height: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:1620): Gtk-CRITICAL **: gtk_widget_get_preferred_width_for_height: assertion 'height >= 0' failed

(mate-panel:1620): GdkPixbuf-CRITICAL **: gdk_pixbuf_get_height: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:1620): Gtk-CRITICAL **: gtk_widget_get_preferred_width_for_height: assertion 'height >= 0' failed
```

Not shure if both issues has the same cause.
##

**Comment by raveit65** [comment]
Confirmed, the commit fixes the warnings, thank you.
To be more precise after adding a launcher from menu or desktop the normal menubar and all launcher freezes. all other applets or wnck applets are still working.

**Comment by monsta** [comment]
> To be more precise after adding a launcher from menu or desktop the normal menubar, all launcher and the panel itself (no context menus) freezes. All other applets or wnck applets are still working.

Yes, exactly that happens in my Debian Testing VM. And not a single warning in ```~/.xsession-errors```...

**Comment by monsta** [comment]
Huh... it's not the launcher that causes it. It's the closing of popup menu. Just right-click on some app in the menu, then make the popup menu go away (doesn't matter how - press Esc, click somewhere outside the menu, etc.).

**Comment by raveit65** [comment]
yes, it's the popup itself,........good finder.

**Comment by raveit65** [comment]
and creating the launcher via drag&drop from menu doesn't freeze the panel

**Comment by raveit65** [comment]
I noticed that the behaviour is much better with mate-panel-1.13.2 and gtk+-3.19.10 .
After adding a starter to panel or desktop from menu right-click option, you only need to move move with the mouse over the menubar items, than the menu closes and you can use the panel starter.

**Comment by monsta** [comment]
Indeed, in 3.20 it seems to work as you described. Same with pressing Esc to close the popup menu. No freezes anymore.

**Comment by raveit65** [comment]
Now, only the menu should close after adding the launcher.

**Comment by monsta** [comment]
Damn, I still can reproduce it with 3.18. :disappointed: 

@flexiondotorg: that's in Ubuntu 16.10, so you'll be able to reproduce it as well.

**Comment by monsta** [comment]
@lukefromdc: by any chance, maybe you know why this and https://github.com/mate-desktop/mate-media/issues/61 are gone with 3.20 but still happen with 3.18...?

**Comment by lukefromdc** [comment]
I don't really know but there were quite a few "underhood" changes in 
GTK 3.20.

**Comment by monsta** [comment]
Agree, and you also made some fixes in mate-panel that went under ```GTK_CHECK_VERSION (3, 19, 0)```... maybe it matters as well.

**Comment by lukefromdc** [comment]
I haven't had this problem show up at all in over a year on my systems, can close the popup menu or add a launcher with no ill effects. I recall stuff like this from when I first switched to MATE but at least not since the 1.10 days if on GTK3 at all.

**Comment by monsta** [comment]
Dammit. Still happens with 3.18.9 for me. Tested with the same method - adding a new launcher from apps/places/system menu.

**Comment by monsta** [comment]
This also happens with GTK+ 3.14 with the same symptoms as with 3.18 (panel freeze on closing the popup menu).

**Comment by monsta** [comment]
Ok, https://github.com/mate-desktop/mate-panel/issues/582 should fix the freeze.

There are still a few issues left that you will find after closing the popup menu (via Esc or clicking somewhere outside of it):
- keyboard focus will be set into the last active app instead of the menu
- menu will remain opened
- can't alt-tab back into menu (no wonder, it's not a normal app window)
- can ctrl-alt-tab into mate-panel, menubar/menu-button will have focus, and you'll be able to use arrow keys to navigate the menu again, or Esc to close it
- can move the mouse to some other category in the menu, that will bring the focus back into the menu as well

This behavior is present in all GTK+3 versions, no matter if the freezing issue is there or not.

**Comment by raveit65** [comment]
I noticed that only open the second popup menu with mozo entries breaks the menu.
Now the menu don't close.
No need to add a launcher to the panel.

**Comment by raveit65** [comment]
@lukefromdc 
This issue is only partial fixed.
Why not renamming it if you merged the PR with closing it?
I told you so in PR.

**Comment by lukefromdc** [comment]
I only intended to close the PR, not the whole issue. The PR is merged as asked, sorry if it messed up the issue itself. 

i can't duplicate the failure of the popup to close with GTK 3.18 (and older libwnck to support it), I suspect that glib version or some other underlying difference from a real GTK 3.18 distro is the reason. In my gtk 3.18 test build it worked perfectly-but mozo was not the launcher I tested either, I just picked one, added it to the panel and the popup both opened and closed as expected. 

Since you are not getting that result there's more work to do so I concur with renaming and reopening.

**Comment by monsta** [comment]
The popup menu for a menuitem closes fine, it's the applications (or settings) menu from the menubar or menu-button applet that won't auto-close.

**Comment by raveit65** [comment]
I tested menu-hacks branch with f26 (gtk+-3.22) and rhel7 (gtk+-3.14).
My results:
- Adding a launcher from menu to the panel/desktop closes the sub-menu (with add to panel entries) and a second click closes the whole menu. Sometimes, especially with some entries from system/systemtools, the second click isn't needed and the whole menu closes fine.
I haven't found out why we see some differences here, but OK for me. 
- Without adding a launcher to the panel/desktop always 2 clicks are needed to close the 2 submenus.
I forgot how this was with gtk+2 ;-)
- But, normally hovering with the mouse over menu-entries open a submenu and if you hover over an other entry the first submenu will close and the next submenu will open.
This does not work if the sub-menu (with add to panel entries) is open. If you open this, a click on desktop/application or another menu-entry is needed to close the submenu.
Hovering with the mouse to another menu-entry will leave the submenu open.

Well, at least this branch improves a lot.
I am fine with merging this like it is. The last point is a very minor issue.

**Comment by lukefromdc** [comment]
Worked for me in my GTK 3.22 test, the problem I had after upgrading (some odd closing behavior and one mouse freeze) but before restarting the panel has not returned, I've been running this for the last day with no post-restart problems at all

**Comment by monsta** [comment]
@raveit65: can you please test 1.16/GTK+2 as well? I was trying to achieve the same behaviour here.

The popup menu (with "add to panel" and stuff) isn't a submenu, it's a separate popup menu. That's how GtkMenu works: it grabs keyboard and mouse focus, so you can't operate on the opened applications menu at the same time. It's by design, and it was so in GTK+2.

**Comment by monsta** [comment]
> Sometimes, especially with some entries from system/systemtools, the second click isn't needed and the whole menu closes fine.

I see it with current git master as well.

I can reproduce it only with GTK+ 3.22, only with menubar applet and only with two-level submenu entries from the categories of settings menu (e.g. settings -> hardware -> sound).
In other cases - GTK+ 3.18, menu-button applet, one-level submenus - I can't reproduce it.

**Comment by monsta** [comment]
I hope you guys tested both menubar and menu-button applets just in case :slightly_smiling_face:

**Comment by raveit65** [comment]
> I hope you guys tested both menubar and menu-button applets just in case 🙂**

I see same behaviour with menubar and mate-main-menu applet, i guess this is what you mean 😃 

> @raveit65: can you please test 1.16/GTK+2 as well? I was trying to achieve the same behaviour here.
(I mean the current 1.16 branch as is, without applying the hacks to it.)

Your right, f24 with gtk+-2 shows the same.
So point 3 from my previous post is obsolete.

**Comment by lukefromdc** [comment]
Submitted a PR for menu-hacks, it's working fine here and in a GTK 3.22/menu-bar test the main menu closes just fine. On adding launchers to the panel the submenus are holding open, which is often a desirable behavior as it makes it easier to add many launchers to the panel from nearby menu entries. Clicking once on desktop, panel, or anything else closes the submenu and menu.

If I open the right-click menu to make the main menu and submenu hold open, clicking back on the same menuitem, any other menuitem, the desktop, the panel, or or anything else closes the right-click menu, leaving the menu and submenu open. The click is captured and does nothing else no matter what was clicked on.

With the menu and submenu holding open and right-click menu closed, clicking on a menu item for a submenu closes the previous submenu and opens the new one, clicking on a menu item for a program closes the menus and opens that program, as expected.

**Comment by monsta** [comment]
Ok... do I have something else to fix to consider this issue closed? :slightly_smiling_face: 

I would really like to make a new 1.18.x release, because people are waiting for some other fixes as well (#494, #504, #572).

**Comment by raveit65** [comment]
> Ok... do I have something else to fix to consider this issue closed? :slightly_smiling_face: 

> I would really like to make a new 1.18.x release, because people are waiting for some other fixes as well (#494, #504, #572).

No, go ahead. :)
We can consider this issue as fixed.
I think all the last fixes should be back port to 1.16 too.
I am not sure if debian maintainer can update their existing packages because of their nonsense policy.

**Comment by monsta** [comment]
I will try to backport these fixes to 1.16 later. No more hurry for Debian though: the release is set to June 17th, and the last deadline for **any** changes in Testing repo is June 9th. Given that each package should always be uploaded to Unstable repo first, and the minimal delay for auto-migration to Testing is 2-3 days, I can say that we're done with it for now. All fixes that didn't make it can be added after the release as patches (new upstream releases will not be allowed into Debian Stable).

That said, the current situation in Debian isn't that bad. Latest 1.16.x versions of Caja, m-s-d, mate-desktop, eom, mate-terminal are there now. Even mate-themes 3.22.11 made it there. :slightly_smiling_face:

**Comment by raveit65** [comment]
I can't understand why a package update from ie. 1.16.2 to 1.16.3 isn't allowed after debian release.
This isn't a major version bump, only a minor version bump for bug fixes.
This makes a distro more buggiest than stable ;-).

**Comment by raveit65** [comment]
PS: I don't think that i will update rhel7 to 1.18 because of switching to gtk+3 during release cycle and i am to lazy for it.
So, rhel7 user will use 1.16 with gtk+-2,  no need to back port gtk+3 fixes to 1.16 from my side.

**Comment by lukefromdc** [comment]
I'm looking forward to Debian Stable releasing so that Unstable unlocks again. Hell I'd probably be better off with Gentoo except for the massive CPU load of compiling a week's worth of general updates after an on-the-road pull of the packages

**Comment by monsta** [comment]
@raveit65: all the fixes are in 1.18 branch now, I think it's ready for the next release. Well, we have #590 left unsolved, but I couldn't find a fix yet.
